### PR TITLE
fix(queue): stabilize heap comparator against transient LocalQueue lookup errors

### LINKED ISSUE #13476: ClusterQueue heap ordering can become inconsistent on transient LocalQueue lookup errors
**What happened**:

When Admission Fair Sharing is enabled, `queueOrderingFunc` calls `CalcLocalQueueFSUsage` during every heap comparison.

If a LocalQueue lookup temporarily fails, that comparison falls back to `baseCmp`. Later comparisons may successfully calculate fair-sharing usage and use fair-sharing ordering instead. As a result, the ordering rule can change between comparisons even though the Workloads have not changed.

This can violate the comparator consistency required by the ClusterQueue heap and potentially leave the heap incorrectly ordered.

**What you expected to happen**:

The heap comparator should use one consistent ordering rule for a given set of Workloads. A transient LocalQueue lookup failure should not cause individual comparisons to switch between fair-sharing and base ordering.

**How to reproduce it (as minimally and precisely as possible)**:

1. Enable Admission Fair Sharing.
2. Create two pending Workloads whose fair-sharing order is opposite to their base priority order:
   - A high-priority Workload in a LocalQueue with higher usage.
   - A low-priority Workload in a LocalQueue with lower usage.
3. Use a client interceptor that returns a transient error for the first LocalQueue lookup and succeeds on subsequent lookups.
4. Compare the same two Workloads twice using `queueOrderingFunc`.
5. The comparison with the transient error uses `baseCmp` and favors the higher-priority Workload.
6. The successful comparison uses fair-sharing usage and favors the lower-usage Workload.

The comparator therefore returns different ordering for unchanged inputs. Using this comparator during heap operations can violate the heap invariant.

**Anything else we need to know?**:

This was noticed while reviewing https://github.com/kubernetes-sigs/kueue/pull/13433#issuecomment-5068404334.

PR #13433 handles the equivalent Snapshot path by precomputing usage and falling back to base ordering for the entire snapshot. The heap path is different because its comparator persists across operations. A safe fix may require stable cached LocalQueue weights or usage inputs and rebuilding affected heap entries when those inputs change, rather than performing fallible API reads inside the comparator.

This should be handled separately from the Snapshot fix in PR #13433.

### PR DESCRIPTION
**Pull request by sanskar-singh-2403**
## What type of PR is this?

/kind bug

## What this PR does / why we need it

When Admission Fair Sharing is enabled, the persistent ClusterQueue heap comparator called `CalcLocalQueueFSUsage` on every comparison, which performed a fallible LocalQueue `Get` to read the fair-sharing weight. On a transient lookup error the comparison fell back to base ordering, while a successful lookup used fair-sharing ordering, so the ordering rule could differ between comparisons for unchanged Workloads and violate the heap invariant.

This caches the LocalQueue weight per ClusterQueue, guarded by the same lock as the heap, and reads the cached weight from the comparator instead of doing an API read. The cache is seeded and kept in sync through the existing `AddLocalQueue` / `UpdateLocalQueue` / `DeleteLocalQueue` hooks; unknown or deleted LocalQueues fall back to the default weight 1.0. On a weight change, with an early return when the weight is unchanged, the heap is rebuilt in O(n) with `heap.Init`: a weight change shifts all of that LocalQueue's workloads at once relative to every other LocalQueue's workloads, so fixing entries one by one is unsafe.

Per the discussion in the issue, this keeps the change scoped to freezing the LocalQueue weights out of the comparator (correctness), and deliberately does not touch the broader reheapify strategy for performance. A debounced workqueue for the reheapify events can be explored separately if hook volume becomes a concern.

## Which issue(s) this PR fixes

Fixes #13476

## Special notes for your reviewer

Includes a deterministic regression test (`TestHeapOrderingStableOnLocalQueueLookupError`) using a client interceptor that fails every LocalQueue lookup, verified to fail against the old comparator and pass with the fix. Also adds `TestHeap_Init`, manager-level reheapify tests covering multi-workload and zero-weight (+Inf usage) transitions, and an integration test that updates `spec.fairSharing.weight` through the Kubernetes API and asserts the new admission order.

cc @YQ-Wang @apullo777 @mimowo

## Release note

```release-note
AFS: Fixed a bug where transient LocalQueue lookup errors in the heap comparator could flip the scheduling order between fair-sharing and priority-based, causing inconsistent workload admission when Admission Fair Sharing was enabled.
```

### CODE REVIEW

**Comment by mimowo** [comment]
cc @kshalot @apullo777 ptal

File: pkg/cache/queue/cluster_queue.go [hunk 1]
```diff
@@ -1018,32 +1036,44 @@ func queueOrderingFunc(
...
 	c.rwm.Lock()
 	defer c.rwm.Unlock()
 	delete(c.localQueuesInClusterQueue, lqKey)
+	delete(c.lqWeights, lqKey)
+}
+
+// UpdateLocalQueueWeight refreshes the cached weight for a LocalQueue and, if it
+// changed, reheapifies that LocalQueue's pending workloads so the heap stays
+// ordered. See Kueue#13476.
+func (c *ClusterQueue) UpdateLocalQueueWeight(lqKey utilqueue.LocalQueueReference, weight float64) {
```
Comment (apullo777) [inline pkg/cache/queue/cluster_queue.go:1218]: There is already a per-LocalQueue reheapify path in `RebuildLocalQueue`, which is called after AFS usage reconciliation via `Manager.RebuildClusterQueue`:

https://github.com/kubernetes-sigs/kueue/blob/030a8029c7ca06f1fd80d07c0a2544416a843bcb/pkg/cache/queue/cluster_queue.go#L477-L485

This adds a second implementation of essentially the same loop.

Could we consolidate both behind a lock-held helper and change `RebuildLocalQueue` to take the full namespaced key, for example:

```
func (c *ClusterQueue) rebuildLocalQueueLocked(
        lqKey utilqueue.LocalQueueReference,
    ) {
        for _, wl := range c.heap.List() {
            if utilqueue.KeyFromWorkload(wl.Obj) == lqKey {
                c.heap.PushOrUpdate(wl)
            }
        }
    }
```

and have both `RebuildLocalQueue` and this call it? Note the two matchers differ today: `RebuildLocalQueue` matches on `Spec.QueueName` only, ignoring the namespace, so it also rebuilds entries of same-named LocalQueues in other namespaces.

File: pkg/cache/queue/cluster_queue.go [hunk 2]
```diff
@@ -186,6 +186,11 @@ type ClusterQueue struct {
 	afsEntryPenalties         *queueafs.AfsEntryPenalties
 	localQueuesInClusterQueue map[utilqueue.LocalQueueReference]bool
 
+	// lqWeights caches each LocalQueue's fair-sharing weight so the heap
+	// comparator can read it without a fallible API call. Missing entries fall
+	// back to weight 1.0. Guarded by rwm. See Kueue#13476.
+	lqWeights map[utilqueue.LocalQueueReference]float64
```
Comment (apullo777) [inline pkg/cache/queue/cluster_queue.go:209]: `localQueuesInClusterQueue` is effectively a set: its values are always `true`, and `hasPendingPenalties` only iterates its keys. With this change, `lqWeights` is populated and deleted at the same LocalQueue lifecycle points.

Could we use `lqWeights` as the membership map as well, where presence means that the LocalQueue belongs to this ClusterQueue, and remove the parallel `map[LocalQueueReference]bool`?

Keeping both maps requires callers to preserve an implicit key-set invariant. `UpdateLocalQueueWeight` does not encode that invariant itself: if it is ever called for an unregistered key, it creates an entry only in `lqWeights`. The current manager paths appear to register the LocalQueue first, so this may not be reachable today, but collapsing the maps would make the invariant structural rather than caller-dependent.

File: pkg/cache/queue/cluster_queue.go [hunk 3]
```diff
@@ -271,9 +276,18 @@ func newClusterQueueImpl(ctx context.Context, client client.Client, wo workload.
...
+			return w
+		}
+		return 1.0
+	}
+	// The comparator reads the sticky workload and cached weights live; safe
+	// because those writes and heap operations all hold rwm.
+	compareFunc := queueOrderingFunc(ctx, getLQWeight, wo, options.fsResWeights, options.enableAdmissionFs, options.afsEntryPenalties, options.afsConsumedResources, sw.matches)
 	// Derive lessFunc from compareFunc for the heap.
 	lessFunc := func(a, b *workload.Info) bool { return compareFunc(a, b) < 0 }
 	// Snapshot sorts without the lock, so it captures the sticky workload once
```
Comment (apullo777) [inline pkg/cache/queue/cluster_queue.go:332]: Non-blocking thought: now that the queue-side ClusterQueue already caches each LocalQueue’s weight, could `buildSnapshotSort` use a snapshot of that cache instead of calling `afs.ResolveLQWeight` again?

That would keep the heap and the pending-workload visibility snapshot on the same weight source and fallback behavior, while avoiding a fresh set of LocalQueue reads each time the snapshot is built.

Since `Snapshot()` sorts after releasing `c.rwm`, we would need to copy the weights under the same RLock used to collect the elements and pass that copy into the comparator. Reading `c.lqWeights` directly during the sort would bring back the same consistency issue we addressed for the sticky workload in #12740.

This can definitely be a follow-up rather than part of this PR. I just think it would be good to track it somewhere, so the heap and the visibility snapshot don’t quietly keep two different weight lookup paths.

File: pkg/cache/queue/manager_test.go [hunk 4]
```diff
@@ -2398,3 +2399,60 @@ func TestDeleteLocalQueue_UnadmittedWorkloads(t *testing.T) {
...
 }
+
+// TestHeadsRespectLocalQueueWeightForPreexistingWorkloads guards the cached-weight
+// seeding order (Kueue#13476): LocalQueues (with their weights) are registered before
+// the ClusterQueue that adopts their pre-existing workloads, matching how the scheduler
+// wires things up. A zero-weight LocalQueue must be the most disadvantaged (+Inf usage),
+// so its workload pops last even though it was created first. If the weight were cached
+// after the workloads are pushed onto the heap, the zero-weight queue would default to
+// weight 1.0 and wrongly pop first.
+func TestHeadsRespectLocalQueueWeightForPreexistingWorkloads(t *testing.T) {
```
Comment (apullo777) [inline pkg/cache/queue/manager_test.go:2586]: The current tests cover the comparator regression and the initial cache-seeding order, but not the state transition introduced by `UpdateLocalQueueWeight`: changing a weight after workloads are already in the heap must reheapify the affected entries.

Could we add a manager-level regression test that queues workloads from two LocalQueues with equal, non-zero consumed usage and weight 1, then changes one LocalQueue’s weight through `Manager.UpdateLocalQueue` and verifies that the next pop reflects the new ordering?

Please keep the entry penalties at zero in this test, so Pop’s existing penalty-driven `rebuildAll` cannot mask a missing reheapify in `UpdateLocalQueueWeight`. This would cover both the manager wiring and the heap repair that this PR relies on.

File: pkg/cache/queue/cluster_queue.go (hunk 1 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go:1218]: Good catch. Consolidated both into a single lock-held helper:

    func (c *ClusterQueue) rebuildLocalQueueLocked(lqKey utilqueue.LocalQueueReference) {
        for _, wl := range c.heap.List() {
            if utilqueue.KeyFromWorkload(wl.Obj) == lqKey {
                c.heap.PushOrUpdate(wl)
            }
        }
    }

Both RebuildLocalQueue and UpdateLocalQueueWeight now call it. I also changed RebuildLocalQueue to take the namespaced LocalQueueReference and threaded that through RebuildClusterQueue and the LocalQueue controller, so it no longer rebuilds same-named LocalQueues in other namespaces. Good spot on that matcher difference.

File: pkg/cache/queue/cluster_queue.go (hunk 2 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go:209]: Done. Removed localQueuesInClusterQueue and made lqWeights the single source of membership (presence in the map means the LocalQueue belongs to this ClusterQueue). hasPendingPenalties now iterates lqWeights, and the populate/delete points are unchanged. This makes the key-set invariant structural rather than caller-dependent, as you suggested.

File: pkg/cache/queue/cluster_queue.go (hunk 3 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go:332]: Agreed this is worth doing but as a follow-up, since as you note Snapshot() sorts after releasing c.rwm, so it would need to copy the weights under the same RLock that collects the elements (reading c.lqWeights during the sort would reintroduce the #12740 consistency issue). I will open a separate issue to track unifying the heap and snapshot weight-lookup paths so they do not diverge. Thanks for flagging it.

File: pkg/cache/queue/manager_test.go (hunk 4 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/manager_test.go:2586]: Added TestUpdateLocalQueueWeightReheapifies: two LocalQueues with equal non-zero usage and weight 1, workloads already on the heap, then Manager.UpdateLocalQueue raises one LocalQueue's weight and the test asserts the next pop reflects the new ordering. Entry penalties are kept at zero so Pop's rebuildAll cannot mask a missing reheapify. I verified it fails if the reheapify in UpdateLocalQueueWeight is removed.

File: pkg/cache/queue/cluster_queue.go [hunk 5]
```diff
@@ -1018,32 +1034,48 @@ func queueOrderingFunc(
...
+// changed, reheapifies that LocalQueue's pending workloads so the heap stays
+// ordered. See Kueue#13476.
+func (c *ClusterQueue) UpdateLocalQueueWeight(lqKey utilqueue.LocalQueueReference, weight float64) {
+	c.rwm.Lock()
+	defer c.rwm.Unlock()
+	if old, ok := c.lqWeights[lqKey]; ok && old == weight {
+		return
+	}
+	c.lqWeights[lqKey] = weight
+	c.rebuildLocalQueueLocked(lqKey)
```
Comment (YQ-Wang) [inline pkg/cache/queue/cluster_queue.go]: Changing a LocalQueue weight changes the ordering of all its workloads at once. Fixing them one by one can still leave the heap invalid; I reproduced `Pop` returning the wrong workload with four entries.

Could we rebuild the heap once with `container/heap.Init` and add a multi-workload regression test?

File: pkg/cache/queue/manager.go [hunk 6]
```diff
@@ -581,12 +583,17 @@ func (m *Manager) UpdateLocalQueue(log logr.Logger, q *kueue.LocalQueue) error {
...
+			newCQ.addLocalQueue(queue.Key(q), afs.LQWeightAsFloat64(q))
 			newCQ.AddFromLocalQueue(qImpl, m.roleTracker, m.customLabels)
-			newCQ.addLocalQueue(queue.Key(q))
 			m.Broadcast()
 		}
 	}
 	qImpl.update(q)
+	// Sync the cached weight with the spec and reheapify if it changed.
+	if newCQ := m.hm.ClusterQueue(q.Spec.ClusterQueue); newCQ != nil {
+		newCQ.UpdateLocalQueueWeight(queue.Key(q), afs.LQWeightAsFloat64(q))
```
Comment (YQ-Wang) [inline pkg/cache/queue/manager.go:595]: Could we add an integration test that updates `spec.fairSharing.weight` through the Kubernetes API and verifies the new admission order?

The current unit test calls the manager directly, so it does not cover the controller update path.

File: pkg/cache/queue/cluster_queue.go [hunk 7]
```diff
@@ -459,14 +473,13 @@ func (c *ClusterQueue) GetNoFitReason(wl workload.Reference) (string, bool) {
 	return string(reason), ok
 }
 
-func (c *ClusterQueue) RebuildLocalQueue(lqName string) {
+// RebuildLocalQueue reheapifies the pending workloads of the given LocalQueue,
+// e.g. after its fair-sharing usage changed. lqKey is the namespaced LocalQueue
+// key so workloads in same-named LocalQueues of other namespaces are not touched.
+func (c *ClusterQueue) RebuildLocalQueue(lqKey utilqueue.LocalQueueReference) {
```
Comment (YQ-Wang) [inline pkg/cache/queue/cluster_queue.go]: This changes an exported parameter from `string` to `LocalQueueReference`, which can break existing callers.

Could we keep the old method as a compatibility wrapper and add a new namespaced method?

File: pkg/cache/queue/cluster_queue.go (hunk 5 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go]: That makes sense. Added Heap.Init (container/heap.Init) and both UpdateLocalQueueWeight and RebuildLocalQueue now rebuild the whole heap in O(n) instead of Fix-ing entries individually; added TestHeap_Init and a 4-workload TestUpdateLocalQueueWeightReheapifiesMultipleWorkloads, both verified to fail against the per-entry approach.

File: pkg/cache/queue/cluster_queue.go (hunk 7 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go]: Reverted to the original RebuildLocalQueue(lqName string) signature (and RebuildClusterQueue/the LQ controller caller with it); 

the whole-heap Init rebuild no longer needs the namespaced key, so there is no longer a breaking signature change to wrap.

File: pkg/cache/queue/manager.go (hunk 6 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/manager.go:595]: Added

**Review by YQ-Wang — APPROVED** [review]
Thanks

File: pkg/cache/queue/cluster_queue.go [hunk 8]
```diff
@@ -464,14 +473,16 @@ func (c *ClusterQueue) GetNoFitReason(wl workload.Reference) (string, bool) {
 	return string(reason), ok
 }
 
+// RebuildLocalQueue reheapifies the pending workloads after a LocalQueue's
+// fair-sharing usage changed. A usage change reorders all of the LocalQueue's
```
Comment (sohankunkerkar) [inline pkg/cache/queue/cluster_queue.go]: A usage change doesn't reorder workloads within the same LQ; they all share the same usage, so their relative order is unchanged. A usage change shifts them relative to other LQs' workloads, which is a per-item move.

File: pkg/cache/queue/manager_test.go [hunk 9]
```diff
@@ -2413,3 +2399,205 @@ func TestDeleteLocalQueue_UnadmittedWorkloads(t *testing.T) {
...
+// TestUpdateLocalQueueWeightReheapifiesMultipleWorkloads guards that a weight
+// change reorders every affected workload at once (Kueue#13476). Two LocalQueues
+// hold two workloads each with equal, non-zero consumed usage and weight 1, so
+// they order by creation time (a1, b1, a2, b2). Raising lq-a's weight halves its
+// fair-sharing usage, so both lq-a workloads must move ahead of both lq-b ones.
+// Fixing the moved entries one by one can leave the heap invariant broken and pop
+// the wrong workload; a single whole-heap rebuild keeps every pop correct. Entry
+// penalties are left at zero so Pop's penalty-driven rebuildAll cannot mask a
+// missing reheapify in UpdateLocalQueueWeight.
+func TestUpdateLocalQueueWeightReheapifiesMultipleWorkloads(t *testing.T) {
```
Comment (sohankunkerkar) [inline pkg/cache/queue/manager_test.go:2716]: nit: could we add a test case for UpdateLocalQueueWeight with weight 0? It exercises the +Inf fallback in CalculateUsage through the runtime update path.

File: pkg/cache/queue/cluster_queue.go [hunk 10]
```diff
@@ -464,14 +473,16 @@ func (c *ClusterQueue) GetNoFitReason(wl workload.Reference) (string, bool) {
...
+// workloads at once, so the whole heap invariant is re-established in O(n) with
+// heap.Init rather than fixing entries one by one (which is unsafe when several
+// items move together). lqName identifies the LocalQueue whose reconcile
+// triggered the rebuild; it is kept for API compatibility and observability.
 func (c *ClusterQueue) RebuildLocalQueue(lqName string) {
```
Comment (sohankunkerkar) [inline pkg/cache/queue/cluster_queue.go]: Is there any plan to log the lqName?

**Review by sohankunkerkar — COMMENTED** [review]
Thanks for the PR. I have a few nits.

/release-note-edit
```release-note
AFS: Fixed a bug where transient LocalQueue lookup errors in the heap comparator could flip the scheduling order between fair-sharing and priority-based, causing inconsistent workload admission when Admission Fair Sharing was enabled.
```

**Comment by sohankunkerkar** [comment]
@sanskar-singh-2403 could you add the release-note  in the github PR description?

File: pkg/cache/queue/cluster_queue.go (hunk 8 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go]: You are right, corrected the wording. The comment now says workloads in the same LocalQueue keep their relative order and all shift at once relative to other LocalQueues' workloads; since heap.Fix assumes a single element moved, fixing them one by one can leave the heap invalid, so heap.Init rebuilds the whole invariant in O(n).

File: pkg/cache/queue/manager_test.go (hunk 9 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/manager_test.go:2716]: Added TestUpdateLocalQueueWeightZeroReheapifies: two LocalQueues with equal non-zero usage and weight 1, then lq-a's weight is dropped to 0 through Manager.UpdateLocalQueue, which makes its usage +Inf via CalculateUsage's non-positive-weight fallback, so its workload pops last. Verified it fails if the reheapify is removed.

File: pkg/cache/queue/cluster_queue.go (hunk 10 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go]: Good idea, done. Threaded a logr.Logger through RebuildLocalQueue and Manager.RebuildClusterQueue (from the reconcile's context logger) and it now logs at V(3): "Rebuilding heap after LocalQueue fair-sharing usage change" with the clusterQueue and localQueue names.

**Comment by sanskar-singh-2403** [comment]
> @sanskar-singh-2403 could you add the release-note in the github PR description?

done, added the same one as yours

**Comment by sanskar-singh-2403** [comment]
hey @mimowo @sohankunkerkar @YQ-Wang @apullo777 can we review this ASAP as it keeps getting merge conflicts, Thanks!

**Review by YQ-Wang — APPROVED** [review]

File: pkg/cache/queue/manager_test.go [hunk 11]
```diff
@@ -2574,3 +2575,275 @@ func TestAddOrUpdateWorkloadCarriesLastAssignment(t *testing.T) {
...
+	}
+	for _, wl := range []*kueue.Workload{wlA1, wlB1, wlA2, wlB2} {
+		if err := manager.AddOrUpdateWorkload(log, wl); err != nil {
+			t.Fatalf("Failed adding workload %s: %v", wl.Name, err)
+		}
+	}
+
+	// Raise lq-a's weight so its usage (4/2) drops below lq-b's (4/1); both
+	// lq-a workloads must now pop before both lq-b workloads.
+	lqA.Spec.FairSharing.Weight = new(resource.MustParse("2"))
```
Comment (apullo777) [inline pkg/cache/queue/manager_test.go]: This one still passes if we remove `c.heap.Init()` entirely. I think it's because `wl-a1` is already the heap root, so raising `lq-a`'s weight doesn't require the head to move; the later pops end up repairing the rest.

Could we flip this to `lq-b` instead?

```go
lqB.Spec.FairSharing.Weight = new(resource.MustParse("2"))
```

and expect `b1, b2, a1, a2`? I tried that locally: it passes with the current code and fails when the reheapify is removed.

One small note: I also couldn't get the new tests to fail against the previous per-entry `PushOrUpdate` loop, so they don't seem to pin the `Init` vs per-entry distinction yet. Not blocking for me; the change above at least makes this a real regression test for losing the reheapify.

File: pkg/cache/queue/cluster_queue.go
Comment (apullo777) [inline pkg/cache/queue/cluster_queue.go]: One nearby thing I noticed: `rebuildAll()` still has the same batch-reorder pattern that motivated `heap.Init()` here. A pending penalty can change the ordering of multiple workloads from the same LocalQueue at once, while this loop re-fixes them one by one.

I reproduced a case where that leaves the heap invalid. Would it make sense to make this just:

```go
func (c *ClusterQueue) rebuildAll() {
    c.heap.Init()
}
```

It's pre-existing, so I'm also fine with a follow-up if we would rather keep this PR scoped.

**Review by apullo777 — COMMENTED** [review]
Thanks for working through the earlier comments. The cached-weight shape looks good to me now :)

I found one remaining gap in the multi-workload test and left a small fix inline. The `rebuildAll` note is pre-existing and non-blocking, so a follow-up is fine there.

File: pkg/cache/queue/manager_test.go (hunk 11 above)
Comment (sanskar-singh-2403) [reply pkg/cache/queue/manager_test.go]: Good catch, the old assertion didn't force the head to move. Flipped it to raise lq-b's weight and expect b1, b2, a1, a2, since lq-b's first workload isn't the heap root, so the very first pop now depends on the reheapify. Confirmed it passes with the fix and fails when heap.Init() is removed from UpdateLocalQueueWeight. The deterministic TestHeap_Init already pins the Init-vs-per-entry distinction at the heap layer.

File: pkg/cache/queue/cluster_queue.go
Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go]: Agreed, rebuildAll had the same batch-reorder problem. Changed it to just c.heap.Init() and added a comment explaining that a pending penalty shifts several workloads from the same LocalQueue at once, so the per-entry PushOrUpdate loop can leave the heap invalid. Folded it into this PR since it's the identical one-line pattern.

**Comment by apullo777** [comment]
/lgtm

LGTM from my side. Leaving the rest to the maintainers.

**Comment by sanskar-singh-2403** [comment]
@tenzen-y @mimowo requesting your reviews here, Thanks!

File: pkg/cache/queue/cluster_queue.go [hunk 12]
```diff
@@ -498,14 +512,19 @@ func (c *ClusterQueue) GetNoFitReason(wl workload.Reference) (string, bool) {
...
-func (c *ClusterQueue) RebuildLocalQueue(lqName string) {
+// RebuildLocalQueue reheapifies the pending workloads after a LocalQueue's
+// fair-sharing usage changed. Workloads in the same LocalQueue share that usage,
+// so their order relative to each other is unchanged; what changes is their
+// position relative to every other LocalQueue's workloads, and all of them shift
+// at once. heap.Fix assumes a single element moved while the rest of the heap is
+// valid, so fixing these entries one by one can leave the heap invalid; instead
+// the whole invariant is re-established in O(n) with heap.Init. lqName identifies
+// the LocalQueue whose reconcile triggered the rebuild.
+func (c *ClusterQueue) RebuildLocalQueue(log logr.Logger, lqName string) {
```
Comment (sohankunkerkar) [inline pkg/cache/queue/cluster_queue.go]: nit: lqName is now only used for logging since heap.Init() rebuilds the entire heap, not just the named LocalQueue's entries. The function name reads like it's scoped to one LocalQueue. Consider renaming to `RebuildHeap` or moving the lqName log context to the caller. Fine as a follow-up.

Comment (sanskar-singh-2403) [reply pkg/cache/queue/cluster_queue.go]: That makes sense. Renamed RebuildLocalQueue to RebuildHeap and updated the doc comment. I kept the lqName param since it's still useful in the V(3) log line to record which LocalQueue's reconcile triggered the rebuild, and it matches the sibling methods that take a logger plus an identifying name. RebuildClusterQueue keeps its name since it does resolve a specific ClusterQueue.

**Review by sohankunkerkar — COMMENTED** [review]
Thanks for working on this!
/approve
/lgtm
